### DISCUSSION TITLE
Prior Art

### DISCUSSION THREAD

**Discussion by emiltin** [rsmp-nordic/rsmp_sxl_vms#3] [Ideas]
What existing examples of SXLs for Variable Messages Signs do we know of, which can be used as inspiration?

**Comment by emiltin** [comment]
SXL for Variable Message Signs used by the Swedish Trafik Administration (Trafikverket):
https://gist.github.com/emiltin/aeb415fb5733fbebdfd484a8bb4d7991

SXL for Variable Message Signs used by the the City of Copenhagen  (Københavns Kommune):
https://gist.github.com/emiltin/ef44f80e161cbe47e213201827ac1f23

**Comment by erik-van-dijk** [comment]
Although it's just a comment for the SXL for Variable Message Signs used by the the City of Copenhagen, it might me good to know that slot 255 was agreed to be used for the image to be shown when connection to the central system was lost. So maybe for a future version extend the comments for S0101 and M0101 with "255=Used at communication failure to the central system"

**Reply by emiltin** [reply]
Thanks. We could also consider providing a way to explicitly control what's shown when connection is lost, and perhaps how long to wait before swiching.

**Comment by emiltin** [comment]
Danish Road Authority, guide on VMSes:
https://www.vejdirektoratet.dk/sites/default/files/2023-07/Brug%20af%20variable%20vejtavler%20og%20vognbanesignaler.pdf


This document (Append 3, library of Variable Message Sign content) from the Danish Road Authority list approved text messages:

https://prodstoragehoeringspo.blob.core.windows.net/02536a22-7245-4c25-8772-7824f26556b9/Bilag%203%20Visningsbibliotek%20for%20variable%20tavler.pdf

**Comment by otterdahl** [comment]
Before RSMP were deployed at Trafikverket, another standard called Sign-XML-DA were used. Sign-XML-DA  is an adaptation of OPC XML-DA which used XML and SOAP.

## Sign types

Sign-XML-DA is designed to be able to control and supervise all types of road signs. In order to assign attributes to the right type of sign, a definition of the different types of signs is provided in  Sign-XML-DA: 

* Sign: A sign object with attributes that apply to all types of signs 
* Graphic sign:  Sign object referring to all signs that use diodes to display messages 
* Fixed sign: Sign object referring to all types of signs that displays any number of  static messages 
* Fullgraphic sign: Sign where any character or symbol can be displayed in one or more  colours 
* Textsign Sign: object with attributes that apply to all types of signs capable of displaying text messages 
* Variable sign: Sign that typically has three different fixed messages that can be  displayed 
* Textsign Matrix: Sign that displays a given number of rows with a fixed number of characters per row 
* Textsign – Rows: Sign that displays a given number of rows with a variable number of  characters depending on font and font size

## Message types

### Sign

- AlarmWrongMessage (Boolean)
- AlarmLocalControlTimeout (Boolean)
- AlarmCommFailure (Boolean)
- AlarmPowerfailure (Boolean)
- AlarmPowersurgeRestart (Boolean) 
- AlarmUPS (Boolean)
- AlarmWigWagError (Boolean)
- ResetAlarmPowerFailure (Boolean)
- SetActivationTime (DateTime)
- SetActiveMessage (unsignedByte)
- SetControl (unsignedByte)
- SetDateTime (DateTime)
- SetDefaultMessage (unsignedByte)
- SetLocalControlTime (unsignedByte)
- SetSignLocation (ArrayOfString)
- SetMessageInterval (unsignedByte)
- StatusActiveMessage (unsignedByte)
- StatusControl (unsignedByte)
- StatusSign (unsignedByte)
- SignWidth (long)
- SignHeight (long)
- SignType (unsignedByte)
- StatusActualMessage (unsignedByte)

### Graphic sign

- AlarmDiodeModuleError (Boolean)
- SetDiodeErrorThreshold (Long)
- AlarmModuleError (Boolean)
- AlarmTempHigh (Boolean)
- AlarmTempLow (Boolean)
- AlarmDoorOpen (Boolean)
- StatusBrightness (UnsignedByte)
- SetBrightness (UnsignedByte)
- SignModulesHorisontal (Short)
- SignModulesVertical (Short)
- ImageType (String)
- Colours (Byte)
- ModulePixelsHorisontal (Short)
- ModulePixelsVertical (Short)
- SignPixelsHorisontal (Long)
- SignPixelsVertical (Long)

### Fullgraphic sign

- SetMessage_x (ArrayOfAny) 

### Textsign 

- SignRows (Long)
- SetMessage_x (ArrayOfStrings)
- CharactersPerRow (Short)
- Font (String)

### Fixed sign

- SetIllumination (Boolean)
- StatusIllumination (Boolean)
- AlarmIlluminationError (Boolean) 
- AlarmPositionUnknown (Boolean)

### Variable sign

- SetActiveMessage (unsignedByte)
- SetMessage_x (ArrayOfAny)
- AlarmMotorFailure (Boolean)


### Example: SetMessage_x (ArrayOfAny) 
The operation is used to store messages in the server/portalserver. The MTOM method is used to send  images (.bmp images) to the server. The images are stored locally and linked to the message internally in the server. If multiple images are sent to the server linked to the same message the server shifts between images with a set interval. If a SetMessage without an attachment is sent to this type of sign
the server responds with an error (E_FAIL). BMP Version 3 (Windows NT) with 32 bits per pixel is used to pass images. No RLE compression is used. MTOM is soley responsible for compressing the images during transfer. The sign can store 50 different messages which mean that x can take any value between 1 and 50. 

The client sends a message containing a dynamic function, a MessageID and a description using three strings. 

There are three dynamic functions to use: 

- 0: No dynamic image 
- 1: The message flashes with a fixed interval 
- 2: The sign switches between two messages (fixed interval) 

The interval between message changes is set by the client using the SetMessageInterval method. The client can read a true representation of the displayed message (including diode errors) using the  StatusActualMessage item or perform a read which returns the stored image that is displayed according to the server/portalserver and the Message_x.value. In this manner it is possible to check what the sign is displaying at the moment.
